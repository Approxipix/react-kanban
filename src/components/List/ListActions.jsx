import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { deleteList } from '../../redux/listReducer/actions';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome/index";
import styled from 'styled-components'
import ClickOutside from "../ClickOutside";

const Wrapper = styled.div`
  position: absolute;
  top: 10%;
  left: 60%;
  width: 10rem;
  background-color: #F4F5F7;
  border-radius: .2rem;
  box-shadow: rgba(0, 0, 0, 0.3) .125rem .125rem .5rem;
  z-index: 10;
`;

const Header = styled.div`
  position: relative;
  width: 100%;
  text-align: center;
  border-bottom: 1px solid rgba(9,30,66,.13);
`;

const Close = styled.button`
  position: absolute;
  top: .5rem;
  right: 0;
  color: #6b778c;
  transition: color .1s ease-in;
  cursor: pointer;
  &:hover {
    color: #42526e;
  }
`;

const Title = styled.h3`
  padding: .75rem .5rem;
  font-size: .9rem;
  color: #6b778c;
  font-weight: normal;
`;


const Action = styled.div`
  padding: .5rem 0;
`;

const Button = styled.button`
  width: 100%;
  padding: .5rem;
  font-size: .875rem;
  color: #40424b;
  font-weight: 600;
  text-align: left; 
  transition: all .2s ease-in;
  cursor: pointer;
  &:hover {
    color: #fff;
    background-color: #2E7EAF;
  }
`;

class ListActions extends Component {
  deleteList = () => {
    const { listId, boardId } = this.props;
    this.props.actions.deleteList({
      boardId: boardId,
      listId: listId,
    });
    this.props.toggleOpened();
  };

  render = () => {
    return (
      <ClickOutside toggleOpened={this.props.toggleOpened}>
        <Wrapper>
          <Header>
            <Title>
              List actions
            </Title>
            <Close onClick={() => this.props.toggleOpened()}>
              <FontAwesomeIcon icon="times" />
            </Close>
          </Header>
          <Action>
            <Button onClick={() => this.deleteList()}>
              Delete List
            </Button>
          </Action>
        </Wrapper>
      </ClickOutside>
    )
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      deleteList: deleteList,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(ListActions);
