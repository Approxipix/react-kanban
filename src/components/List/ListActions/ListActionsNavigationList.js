const ListActionsNavigationList = [
  {
    title: 'Delete List',
    component: 'ListDelete',
  },
];

export default ListActionsNavigationList
