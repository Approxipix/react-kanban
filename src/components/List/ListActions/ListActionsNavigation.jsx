import React, { Component } from 'react';
import ListActionsNavigationList from "./ListActionsNavigationList";
import styled from 'styled-components'

const List = styled.ul`
`;

const Item = styled.li`
  cursor: pointer;
`;

const Link = styled.a`
  display: block;
  padding: .3rem .75rem;
  font-weight: 700;
  transition: all .2s ease-in;
  &:hover {
    color: #fff;
    background-color: #2E7EAF;
  }
`;
Link.displayName = 'Link';

class ListActionsNavigation extends Component {
  constructor(props) {
    super(props);
    this.state = {
      navigation: ListActionsNavigationList,
    }
  }

  changeView = (e, item) => {
    e.preventDefault();
    this.props.changeView(item.component, item.title)
  };

  render() {
    const { navigation } = this.state;
    return (
      <List>
        {navigation.map((item, index) => (
          <Item key={index} >
            <Link onClick={(e) => this.changeView(e, item)}>
              {item.title}
            </Link>
          </Item>
        ))}
      </List>
    );
  }
}

export default ListActionsNavigation;
