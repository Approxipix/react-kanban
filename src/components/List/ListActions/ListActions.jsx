import React, { Component } from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import ClickOutside from "../../ClickOutside/ClickOutside";
import ListActionsNavigation from './ListActionsNavigation';
import ListDeleteAllCards from '../ListDeleteAllCards';
import ListMoveAllCards from '../ListMoveAllCards';
import ListDelete from '../ListDelete';
import ListMove from '../ListMove';
import styled from 'styled-components';

const Wrapper = styled.div`
  position: absolute;
  top: 2rem;
  left: calc(100% - 2rem);
  width: 17rem;
  background-color: #F4F5F7;
  border-radius: .2rem;
  box-shadow: rgba(0, 0, 0, 0.3) .125rem .125rem .5rem;
  z-index: 10;
`;

const Container = styled.div`
  padding: 0 .5rem;
`;

const Header = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: .4rem;
  border-bottom: 1px solid rgba(9,30,66,.13);
`;

const Body = styled.div`
   padding: .5rem 0;
`;

const Title = styled.h3`
  font-size: 1em;
  color: #6b778c;
`;

const Button = styled.button`
  color: #6b778c;
  font-size: 1rem;
  transition: color .1s ease-in;
  &:hover {
    color: #42526e;
  }
`;

class ListActions extends Component {
  constructor(props) {
    super(props);
    this.state = {
      actionView: null,
      actionTitle: 'List Actions',
    }
  }

  changeView = (component, title) => {
    this.setState({
      actionView: component,
      actionTitle: title,
    });
  };

  actionView = () => {
    const { boardId, listId } = this.props;
    const { actionView } = this.state;
    switch (actionView) {
      case ('ListMove'):
        return (
          <Container>
            <ListMove boardId={boardId} listId={listId} toggleOpened={this.props.toggleOpened} />
          </Container>
        );
      case ('ListMoveAllCards'):
        return (
          <Container>
            <ListMoveAllCards boardId={boardId} listId={listId} toggleOpened={this.props.toggleOpened} />
          </Container>
        );
      case ('ListDeleteAllCards'):
        return (
          <Container>
            <ListDeleteAllCards listId={listId} toggleOpened={this.props.toggleOpened} />
          </Container>
        );
      case ('ListDelete'):
        return (
          <Container>
            <ListDelete boardId={boardId} listId={listId}/>
          </Container>
        );
      default:
        return <ListActionsNavigation changeView={this.changeView}/>;
    }
  };

  render = () => {
    const { actionView, actionTitle } = this.state;

    return (
      <ClickOutside action={'mousedown'} toggleOpened={this.props.toggleOpened}>
        <Wrapper>
          <Header>
            <Button disabled={!actionView} onClick={() => this.changeView(null, 'List Actions')}>
              {!!actionView && <FontAwesomeIcon icon="chevron-left" />}
            </Button>
            <Title>
              {actionTitle}
            </Title>
            <Button onClick={() => this.props.toggleOpened}>
              <FontAwesomeIcon icon="times" />
            </Button>
          </Header>
          <Body>
            {this.actionView()}
          </Body>
        </Wrapper>
      </ClickOutside>
    )
  };
}

ListActions.defaultProps = {
  boardId: '',
  listId: '',
  toggleOpened: () => {},
};

ListActions.propTypes = {
  boardId: PropTypes.string.isRequired,
  listId: PropTypes.string.isRequired,
  toggleOpened: PropTypes.func.isRequired,
};

export default ListActions;
