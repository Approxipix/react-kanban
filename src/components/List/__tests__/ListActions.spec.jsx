import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import ListActions from '../ListActions/ListActions';


describe('<ListActions>', () => {
  let store;
  let component;

  beforeEach(() => {
    const state = {
      actionView: null,
      actionTitle: 'List Actions',
    };

    component = shallow(<ListActions store={store}  />);
    component.setState({ ...state });
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if changeView sets actionView and actionTitle correctly', () => {
    component.instance().changeView('TestComponentView', 'Test Component');
    expect(component.instance().state.actionTitle).toEqual('Test Component');
    expect(component.instance().state.actionView).toEqual('TestComponentView');
  });

  it('it should render fontawesome icon if sidebarView was selected', () => {
    component.instance().changeView('TestComponentView', 'Test Component');

    expect(toJson(component)).toMatchSnapshot();
  });
});
