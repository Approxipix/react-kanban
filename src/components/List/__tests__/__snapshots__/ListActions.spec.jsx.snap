// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListActions> it should render fontawesome icon if sidebarView was selected 1`] = `
<ClickOutside
  action="mousedown"
  toggleOpened={[Function]}
>
  <styled.div>
    <styled.div>
      <styled.button
        disabled={false}
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon="chevron-left"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </styled.button>
      <styled.h3>
        Test Component
      </styled.h3>
      <styled.button
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon="times"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </styled.button>
    </styled.div>
    <styled.div>
      <ListActionsNavigation
        changeView={[Function]}
      />
    </styled.div>
  </styled.div>
</ClickOutside>
`;

exports[`<ListActions> should render with given state from Redux store 1`] = `
<ClickOutside
  action="mousedown"
  toggleOpened={[Function]}
>
  <styled.div>
    <styled.div>
      <styled.button
        disabled={true}
        onClick={[Function]}
      />
      <styled.h3>
        List Actions
      </styled.h3>
      <styled.button
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon="times"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </styled.button>
    </styled.div>
    <styled.div>
      <ListActionsNavigation
        changeView={[Function]}
      />
    </styled.div>
  </styled.div>
</ClickOutside>
`;
